src/main/java/com/algorand/algosdk/logic/Logic.java

```diff
@@ -0,0 +1,224 @@
+                );
+            } catch (UnsupportedEncodingException ex) {
+                throw new IllegalStateException("langspec opening error");
+            }
+
+            Gson g = new GsonBuilder().create();
+
+            langSpec = g.fromJson(reader, LangSpec.class);
+            reader.close();
+        }
+
+        VariantResult result = Uvariant.parse(program);
+        int vlen = result.length;
+        if (vlen <= 0) {
+            throw new IllegalArgumentException("version parsing error");
+        }
+
+        int version = result.value;
+        if (version > langSpec.EvalMaxVersion) {
+            throw new IllegalArgumentException("unsupported version");
+        }
+
+        if (args == null) {
+            args = new ArrayList<byte[]>();
+        }
+
+        int cost = 0;
+        int length = program.length;
+        for (int i = 0; i < args.size(); i++) {
+            length += args.get(i).length;
+        }
+
+        if (length > MAX_LENGTH) {
+            throw new IllegalArgumentException("program too long");
+        }
+
+        if (opcodes == null) {
+            opcodes = new Operation[256];
+            for (int i = 0; i < langSpec.Ops.length; i++) {
+                Operation op = langSpec.Ops[i];
+                opcodes[op.Opcode] = op;
+            }
+        }
+
+        int pc = vlen;
+        while (pc < program.length) {
+            int opcode = program[pc] & 0xFF;
+            Operation op = opcodes[opcode];
+            if (op == null) {
+                throw new IllegalArgumentException("invalid instruction");
+            }
+
+            cost += op.Cost;
+            int size = op.Size;
+            if (size == 0) {
+                switch (op.Opcode) {
+                    case INTCBLOCK_OPCODE:
+                        size = checkIntConstBlock(program, pc);
+                        break;
+                    case BYTECBLOCK_OPCODE:
+                        size = checkByteConstBlock(program, pc);
+                        break;
+                    default:
+                        throw new IllegalArgumentException("invalid instruction");
+                }
+            }
+            pc += size;
+        }
+
+        if (cost > MAX_COST) {
+            throw new IllegalArgumentException("program too costly to run");
+        }
+
+        return true;
+    }
+
+    static int checkIntConstBlock(byte[] program, int pc) {
+        int size = 1;
+        VariantResult result = Uvariant.parse(Arrays.copyOfRange(program, pc + size, program.length));
+        if (result.length <= 0) {
+            throw new IllegalArgumentException(
+                String.format("could not decode int const block at pc=%d", pc)
+            );
+        }
+        size += result.length;
+        int numInts = result.value;
+        for (int i = 0; i < numInts; i++) {
+            if (pc + size >= program.length) {
+                throw new IllegalArgumentException("intcblock ran past end of program");
```

Comment:
Consider providing a more explicit error message. I believe that the dev who calls it doesn't aware of `intcblock`

Reply:
done